java/vector/src/test/java/org/apache/arrow/vector/TestDictionaryVector.java

```diff
@@ -142,4 +148,194 @@ public void testEncodeLargeVector() {
+
+      writer.setPosition(3);
+      writer.startList();
+      writer.integer().writeInt(30);
+      writer.integer().writeInt(40);
+      writer.integer().writeInt(50);
+      writer.endList();
+
+      writer.setPosition(4);
+      writer.startList();
+      writer.integer().writeInt(30);
+      writer.integer().writeInt(40);
+      writer.integer().writeInt(50);
+      writer.endList();
+
+      writer.setPosition(5);
+      writer.startList();
+      writer.integer().writeInt(10);
+      writer.integer().writeInt(20);
+      writer.endList();
+
+      writer.setValueCount(6);
+
+      UnionListWriter dictWriter = dictionaryVector.getWriter();
+      dictWriter.allocate();
+
+      dictWriter.setPosition(0);
+      dictWriter.startList();
+      dictWriter.integer().writeInt(10);
+      dictWriter.integer().writeInt(20);
+      dictWriter.endList();
+
+      dictWriter.setPosition(1);
+      dictWriter.startList();
+      dictWriter.integer().writeInt(30);
+      dictWriter.integer().writeInt(40);
+      dictWriter.integer().writeInt(50);
+      dictWriter.endList();
+
+      dictWriter.setValueCount(2);
+
+      Dictionary dictionary = new Dictionary(dictionaryVector, new DictionaryEncoding(1L, false, null));
+
+      try (final ValueVector encoded = (FieldVector) DictionaryEncoder.encode(vector, dictionary)) {
+        // verify indices
+        assertEquals(IntVector.class, encoded.getClass());
+
+        IntVector index = ((IntVector)encoded);
+        assertEquals(6, index.getValueCount());
+        assertEquals(0, index.get(0));
+        assertEquals(0, index.get(1));
+        assertEquals(0, index.get(2));
+        assertEquals(1, index.get(3));
+        assertEquals(1, index.get(4));
+        assertEquals(0, index.get(5));
+
+        // now run through the decoder and verify we get the original back
+        try (ValueVector decoded = DictionaryEncoder.decode(encoded, dictionary)) {
+          assertEquals(vector.getClass(), decoded.getClass());
+          assertEquals(vector.getValueCount(), decoded.getValueCount());
+          for (int i = 0; i < 5; i++) {
+            assertEquals(vector.getObject(i), decoded.getObject(i));
+          }
+        }
+      }
+    }
+  }
+
+  @Test
+  public void testEncodeStruct() {
+    // Create a new value vector
+    try (final StructVector vector = StructVector.empty("vector", allocator);
+        final StructVector dictionaryVector = StructVector.empty("dict", allocator);) {
+      vector.addOrGet("f0", FieldType.nullable(new ArrowType.Int(32, true)), IntVector.class);
+      vector.addOrGet("f1", FieldType.nullable(new ArrowType.Int(64, true)), BigIntVector.class);
+      dictionaryVector.addOrGet("f0", FieldType.nullable(new ArrowType.Int(32, true)), IntVector.class);
+      dictionaryVector.addOrGet("f1", FieldType.nullable(new ArrowType.Int(64, true)), BigIntVector.class);
+
+      NullableStructWriter writer = vector.getWriter();
+      writer.allocate();
+
+      writer.start();
```

Comment:
I think this test would be clearer if you added helper methods for addStruct so it is clearer to see which entries are actually equal

Reply:
Sure, fixed.